src/ComputeSharp.D2D1/Shaders/Interop/Effects/PixelShaderEffect.cs

```diff
@@ -0,0 +1,462 @@
+using UnmanagedCallersOnlyAttribute = ComputeSharp.NetStandard.System.Runtime.InteropServices.UnmanagedCallersOnlyAttribute;
+#endif
+
+namespace ComputeSharp.D2D1.Interop.Effects;
+
+/// <summary>
+/// A simple <see cref="ID2D1EffectImpl"/> and <see cref="ID2D1DrawTransform"/> implementation for a given pixel shader.
+/// </summary>
+internal unsafe partial struct PixelShaderEffect
+{
+    /// <summary>
+    /// A wrapper for an effect factory.
+    /// </summary>
+    /// <param name="effectImpl">The resulting effect factory.</param>
+    /// <returns>The <c>HRESULT</c> for the operation.</returns>
+    [UnmanagedFunctionPointer(CallingConvention.StdCall)]
+    private delegate int FactoryDelegate(IUnknown** effectImpl);
+
+#if !NET6_0_OR_GREATER
+    /// <inheritdoc cref="QueryInterface"/>
+    [UnmanagedFunctionPointer(CallingConvention.StdCall)]
+    private delegate int QueryInterfaceDelegate(PixelShaderEffect* @this, Guid* riid, void** ppvObject);
+
+    /// <inheritdoc cref="AddRef"/>
+    [UnmanagedFunctionPointer(CallingConvention.StdCall)]
+    private delegate uint AddRefDelegate(PixelShaderEffect* @this);
+
+    /// <inheritdoc cref="Release"/>
+    [UnmanagedFunctionPointer(CallingConvention.StdCall)]
+    private delegate uint ReleaseDelegate(PixelShaderEffect* @this);
+
+    /// <inheritdoc cref="GetConstantBuffer"/>
+    [UnmanagedFunctionPointer(CallingConvention.StdCall)]
+    public delegate int GetConstantBufferDelegate(IUnknown* effect, byte* data, uint dataSize, uint* actualSize);
+
+    /// <inheritdoc cref="SetConstantBuffer"/>
+    [UnmanagedFunctionPointer(CallingConvention.StdCall)]
+    public delegate int SetConstantBufferDelegate(IUnknown* effect, byte* data, uint dataSize);
+
+    /// <summary>
+    /// A cached <see cref="GetConstantBufferDelegate"/> instance wrapping <see cref="GetConstantBuffer"/>.
+    /// </summary>
+    public static readonly GetConstantBufferDelegate GetConstantBufferWrapper = GetConstantBuffer;
+
+    /// <summary>
+    /// A cached <see cref="SetConstantBufferDelegate"/> instance wrapping <see cref="SetConstantBuffer"/>.
+    /// </summary>
+    public static readonly SetConstantBufferDelegate SetConstantBufferWrapper = SetConstantBuffer;
+#endif
+
+    /// <summary>
+    /// A generic pixel shader implementation.
+    /// </summary>
+    /// <typeparam name="T">The type of shader.</typeparam>
+    public static class For<T>
+        where T : unmanaged, ID2D1PixelShader
+    {
+        /// <summary>
+        /// The <see cref="FactoryDelegate"/> wrapper for the shader factory.
+        /// </summary>
+        private static readonly FactoryDelegate EffectFactory = CreateEffect;
+
+        /// <summary>
+        /// Indicates whether or not initialization has completed.
+        /// </summary>
+        private static bool isInitialized;
+
+        /// <summary>
+        /// The <see cref="Guid"/> for the shader.
+        /// </summary>
+        [FixedAddressValueType]
```

Comment:
this makes it impossible to use PixelShaderEffect in a collectible assembly (that is, an assembly loaded into a collectible AssemblyLoadContext)

<img width="530" alt="image" src="https://user-images.githubusercontent.com/10873410/163920030-e2f7df08-5577-48cb-9ec7-95dd0e5d92a4.png">